### DISCUSSION TITLE
ExpressLRS poor software design

### DISCUSSION THREAD

**Discussion by ElektroBOXua** [ExpressLRS/ExpressLRS#2620] [General]
While working with ExpressLRS, I encountered many problems. Most of them are related to poor software design.
- Adding new features is a pain. For example, there is no unified abstraction for radio drivers. Porting a new driver is a lot of work. In particular, rewriting the entire driver code to comply with ELRS.
- Violation of single responsibility. Many modules are closely interconnected and do more than they need to. For example, FHSS is responsible for many things (domain management, gemini, and even dual-band for now). But ideally, FHSS should only define the logic of frequency hopping. Radio drivers are designed to control two modules simultaneously, which again violates single responsibility.
- Poor modularity. Upgrading functions **this way** looks fundamentally wrong:
```cpp
static inline uint32_t FHSSgetGeminiFreq()
{
    if (FHSSuseDualBand)
    {
        // When using Dual Band there is no need to calculate an offset frequency. Unlike Gemini with 2 frequencies in the same band.
        return FHSSconfigDualBand->freq_start + (FHSSsequence_DualBand[FHSSptr] * freq_spread_DualBand / FREQ_SPREAD_SCALE);
    }
    else
    {
        if (FHSSusePrimaryFreqBand)
        {
            return FHSSGeminiFreq(FHSSsequence[FHSSgetCurrIndex()]);
        }
        else
        {
            return FHSSGeminiFreq(FHSSsequence_DualBand[FHSSgetCurrIndex()]);
        }
    }
}
```

Multiple instances of an FHSS class is the way to go.

Does anyone else feel that way, or it's just me? What design problems are more concerning to you?

**Comment by Linjieqiang** [comment]
You can commit your PR to perfect it. You can you up~!
